docs/design/sql/10-distributed-scan.md

```diff
@@ -0,0 +1,146 @@
+# Distributed Scan
+
+## Overview
+
+The Hazelcast Mustang is a distributed query engine. When an SQL string is submitted for execution, it is converted into a 
+query plan, that contains the tree of relational operators, such as `scan`, `project` and `filter`. Leaf operators of the plan 
+act as data sources for the parent operators. The `scan` operator iterates over the `IMap` data structure. This document 
+explains how the distributed scan is performed.
+
+In this document, we describe the main design points of the scan operators: access path selection, local execution semantics,
+and reaction to cluster reconfiguration.
+
+## 1 Background
+
+The Hazelcast IMDG is a distributed in-memory key-value storage. Data is stored in distributed objects, such as an `IMap`.
+The cluster has a predefined number of partitions, 271 by default. Each partition is stored on a single member, and may have 
+zero, one or more backups on other members. Every distributed object is split across one or more partitions. Partitions may
+migrate between members due to topology change events, such as member leave or join.
+
+The `IMap` is a distributed map. It could be accessed either directly, or through a secondary index. The secondary index is 
+a distributed data structure. Every member stores part of the index that is build for the local entries of the member. 
+
+The goal of the Hazelcast Mustang engine is to pick the proper access method - direct scan or index scan, start execution of 
+the scan on members, and collect the results. In the above sections we describe how it is implemented.   
+
+## 2 Access Path Selection
+
+There are two ways to scan the `IMap` - iterate over the record store directly, or use one of the secondary indexes. During the
+planning stage the proper access method is chosen. 
+
+If there are no indexes on the table, the direct scan is chosen and no further optimization is performed.    
+
+If there are indexes on the table, we analyze the predicate stored in the table of the `TableScan` operator. The entry point
+is `IndexResolver.createIndexScans`. 
+
+First, we split the predicate into conjunctive normal form (CNF). For example, the predicate `a=1 AND b=2` is split into 
+`a=1` and `b=2`, while the predicate `a=1 OR b=2` remains unchanged.
```

Comment:
What if we have an index for `(a, b)`: if we split the predicate like this, will we find the index? If yes, this algorithm isn't described.

Reply:
It is explained in the next paragraphs - we assign every column a list of candidate predicates, then we iterate over all indexes, trying to build the index filter, using column predicates.

Reply:
I guess it's like this:
- input: predicate:  `a=1 AND b=2`, index: `HASH(a, b)`
- we split the predicate into `a=1` and `b=2`
- for each sub-predicate we find an index that can be used. For `a=1` the `HASH(a, b)` index can be used. For `b=1`, no index can be used. We continue with `a=1` only
- when applying `a=1` to the `HASH(a, b)` index we try to apply more predicates. We realize that we can also use the `b=1` condition. And this step is not mentioned in the text, at least I can't find it. I think it happens [in this method](https://github.com/hazelcast/hazelcast/blob/7f16ac358148ce322dd6cc1b44ad3f85adef2d0b/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java#L638).

Reply:
This is explained in the paragraph "Next, we iterate over every index, and try to bind candidates ...". I expanded the explanation a bit: https://github.com/hazelcast/hazelcast/pull/17659/commits/612958c5dffa19fe79cf5b6f45c12db76efc4bc4